client/components/site-selector/index.jsx

```diff
@@ -449,7 +416,53 @@ class SiteSelector extends Component {
 	}
 }
 
-export default connect( ( state ) => {
+const navigateToSite = ( siteId, {
+	allSitesPath,
+	allSitesSingleUser,
+	siteBasePath,
+} ) => ( dispatch, getState ) => {
+	const pathname = getPathnameForSite( siteId );
+	if ( pathname ) {
+		page( pathname );
+	}
+
+	function getPathnameForSite() {
+		const site = getSite( getState(), siteId );
+		debug( 'getPathnameForSite', siteId, site );
+
+		if ( siteId === ALL_SITES ) {
+			// default posts links to /posts/my when possible and /posts when not
+			const postsBase = allSitesSingleUser ? '/posts' : '/posts/my';
+			const path = allSitesPath.replace( /^\/posts\b(\/my)?/, postsBase );
+
+			// There is currently no "all sites" version of the insights page
+			return path.replace( /^\/stats\/insights\/?$/, '/stats/day' );
+		} else if ( siteBasePath ) {
+			return getSiteBasePath( site ) + '/' + site.slug;
+		}
+	}
+
+	function getSiteBasePath( site ) {
+		let path = siteBasePath;
+		const postsBase = ( site.jetpack || site.single_user_site ) ? '/posts' : '/posts/my';
+
+		// Default posts to /posts/my when possible and /posts when not
+		path = path.replace( /^\/posts\b(\/my)?/, postsBase );
+
+		// Default stats to /stats/slug when on a 3rd level post/page summary
+		if ( path.match( /^\/stats\/(post|page)\// ) ) {
+			path = '/stats';
+		}
+
+		if ( path.match( /^\/domains\/manage\// ) ) {
+			path = '/domains/manage';
+		}
+
+		return path;
+	}
+};
```

Comment:
- [ ] Are there precendents of such thunks that never dispatch any action and instead only call `getState` and perform some side effect (in this case, `page` call)?

Note that, in the future, if we add middleware-level routing support (e.g. `dispatch( { type: 'NAVIGATE', path } )`, the above becomes a "normal" thunk which calls `getState` in order to have sufficient data to then `dispatch` a routing request.

- [ ] The remaining question: given its nature, where should this action creator live?